llvm/lib/Target/SPIRV/SPIRVTargetMachine.cpp

```diff
@@ -91,6 +97,88 @@ SPIRVTargetMachine::SPIRVTargetMachine(const Target &T, const Triple &TT,
   setRequiresStructuredCFG(false);
 }
 
+enum AddressSpace {
+  Function = storageClassToAddressSpace(SPIRV::StorageClass::Function),
+  CrossWorkgroup =
+      storageClassToAddressSpace(SPIRV::StorageClass::CrossWorkgroup),
+  UniformConstant =
+      storageClassToAddressSpace(SPIRV::StorageClass::UniformConstant),
+  Workgroup = storageClassToAddressSpace(SPIRV::StorageClass::Workgroup),
+  Generic = storageClassToAddressSpace(SPIRV::StorageClass::Generic)
+};
+
+unsigned SPIRVTargetMachine::getAssumedAddrSpace(const Value *V) const {
+  const auto *LD = dyn_cast<LoadInst>(V);
+  if (!LD)
+    return UINT32_MAX;
+
+  // It must be a load from a pointer to Generic.
+  assert(V->getType()->isPointerTy() &&
+         V->getType()->getPointerAddressSpace() == AddressSpace::Generic);
+
+  const auto *Ptr = LD->getPointerOperand();
+  if (Ptr->getType()->getPointerAddressSpace() != AddressSpace::UniformConstant)
+    return UINT32_MAX;
+  // For a loaded from a pointer to UniformConstant, we can infer CrossWorkgroup
+  // storage, as this could only have been legally initialised with a
+  // CrossWorkgroup (aka device) constant pointer.
+  return AddressSpace::CrossWorkgroup;
+}
+
+std::pair<const Value *, unsigned>
+SPIRVTargetMachine::getPredicatedAddrSpace(const Value *V) const {
+  using namespace PatternMatch;
+
+  if (auto *II = dyn_cast<IntrinsicInst>(V)) {
+    switch (II->getIntrinsicID()) {
+    case Intrinsic::amdgcn_is_shared:
+      return std::pair(II->getArgOperand(0), AddressSpace::Workgroup);
+    case Intrinsic::amdgcn_is_private:
+      return std::pair(II->getArgOperand(0), AddressSpace::Function);
+    default:
+      break;
+    }
+    return std::pair(nullptr, UINT32_MAX);
+  }
+  // Check the global pointer predication based on
+  // (!is_share(p) && !is_private(p)). Note that logic 'and' is commutative and
+  // the order of 'is_shared' and 'is_private' is not significant.
+  Value *Ptr;
+  if (getTargetTriple().getVendor() == Triple::VendorType::AMD &&
+      match(
+          const_cast<Value *>(V),
+          m_c_And(m_Not(m_Intrinsic<Intrinsic::amdgcn_is_shared>(m_Value(Ptr))),
+                m_Not(m_Intrinsic<Intrinsic::amdgcn_is_private>(
+                    m_Deferred(Ptr))))))
```

Comment:
Shouldn't be looking at the amdgcn intrinsics? Surely spirv has its own operations for this?

Reply:
There's AMDGCN flavoured SPIR-V, which'd possibly have these in source; I don't think there's AS predicates in SPIR-V, at least not AFAICS in Clang/LLVM/the spec - happy to add them if they exist, but we'll need both.

Reply:
If I have skimmed SPIRV correctly, it expects invalid addrspacecasts (OpGenericCastToPtrExplicit) to return null. You could implement the same kind of check by looking for icmp ne (addrspacecast x to y), null

Reply:
Neither the BE nor the Translator handle that at the moment, and I suspect it's meant for implementing some specific bit of OpenCL (SYCL?) functionality. We use the non-explicit flavours, and those don't return null (and are diagnosed as illegal if they are illegal per spec). This is probably a good way of implementing the predicates / handling this, so thank you for it. Having said that, I reiterate that we have AMDGCN flavoured SPIR-V where the actual AMDGCN predicates would manifest / make sense.

Reply:
We could do the same thing for amdgpu. We implement addrspacecast with the same operations.

This also reminds me, we should have a valid flag on addrspacecast.

Reply:
```
If I have skimmed SPIRV correctly, it expects invalid addrspacecasts (OpGenericCastToPtrExplicit) to return null. You could implement the same kind of check by looking for icmp ne (addrspacecast x to y), null
```

The semantic described in Langref doesn't allow you to map `addrspacecast` to `OpGenericCastToPtrExplicit` as no check is performed. [OpGenericCastToPtr](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#OpGenericCastToPtr) is what I expect `addrspacecast` to be mapped to.

Reply:
I do think we need to add a poison-if-known-invalid-cast flag to addrspacecast

Reply:
> I do think we need to add a poison-if-known-invalid-cast flag to addrspacecast

+1 (so does SPIR-V but that's another story)

Reply:
> ```
> If I have skimmed SPIRV correctly, it expects invalid addrspacecasts (OpGenericCastToPtrExplicit) to return null. You could implement the same kind of check by looking for icmp ne (addrspacecast x to y), null
> ```
> 
> The semantic described in Langref doesn't allow you to map `addrspacecast` to `OpGenericCastToPtrExplicit` as no check is performed. [OpGenericCastToPtr](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#OpGenericCastToPtr) is what I expect `addrspacecast` to be mapped to.

I think that we just need to implement the AS predicates (`is_local` / `is_private` & friends) atop `OpGenericPtrMemSemantics`; not as part of this patch but in general, orthogonally to adding poison-if-known-invalid-cast.

Reply:
Oh just seeing this comment @AlexVlx

> I think that we just need to implement the AS predicates (is_local / is_private & friends) atop `OpGenericPtrMemSemantics`

is that for AMDGCN or something more general ? If the latter, the spec doesn't offer enough guarantee to do that.

Reply:
I was thinking about generic (general) predicates for SPIR-V. AFAICS the spec says this about `OpGenericPtrMemSemantics`: `Result is a valid [Memory Semantics](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#Memory_Semantics_-id-) which includes mask bits set for the Storage Class for the specific (non-Generic) Storage Class of Pointer`. My interpretation (which could be wrong) is that the bits returned in the mask actually indicate the pointee's AS, so the generic predicates would lower to (handwavium alert) `OpGenericPtrMemSemantics` + bitwise AND.

Reply:
> My interpretation (which could be wrong) is that the bits returned in the mask actually indicate the pointee's AS, so the generic predicates would lower to (handwavium alert) OpGenericPtrMemSemantics + bitwise AND.

The returned value is guaranteed to be a valid combination for the AS but an impl can use the same combination for different AS.

Reply:
I’m not quite sure how to parse this, apologies - what is an implementation in this case? It would be rather odd to have a valid implementation use e.g. setting the WorkGroup bit to denote CrossWorkGroup, would it not? Note I’m only considering SPIR-V, not what a target would decide to lower it to.

Reply:
> what is an implementation in this case?

a tool consuming the SPIR-V module like an opencl driver

> It would be rather odd to have a valid implementation use e.g. setting the WorkGroup bit to denote CrossWorkGroup, would it not?

it is, but it may not make a difference for all platforms (e.g. CPUs don't typically have a dedicated workgroup memory) and checking what you are dealing can be somehow expensive or complex for no clear benefit down the line.